File: 2015/13/rust/src/main.rs
use itertools::Itertools;
use regex::Regex;
use std::collections::{ HashMap, HashSet };
use std::fs::File;
use std::io::{ BufRead, BufReader };

// --- Constants --
const INPUT_FILE: &'static str = "../input/input.txt";
const PARSER_REGEX: &'static str = r"(.+)? would (gain|lose) (\d+) happiness units by sitting next to (.+)\.";

// --- Calculate happiness ---
fn calculate_happiness(persons: &Vec<String>, happiness: &HashMap<(String, String), isize>) -> isize {
    let mut max = std::isize::MIN;
    for c in persons.iter().permutations(persons.len()).unique() {
        let mut h = 0;
        for i in 0 .. c.len() {
            let n1 = c[i];
            let n2 = c[(i + 1) % c.len()];
            h += happiness.get(&(n1.to_string(), n2.to_string())).unwrap();
            h += happiness.get(&(n2.to_string(), n1.to_string())).unwrap();
        }
        max = std::cmp::max(max, h);
    }
    max
}

// --- Main function ---
fn main() -> Result<(), Box<dyn std::error::Error>> {
    let regex = Regex::new(PARSER_REGEX)?;

    let mut persons: HashSet<String> = HashSet::new();
    let mut happiness: HashMap<(String, String), isize> = HashMap::new();

    // --- Read and parse the input file ---
    let file = File::open(INPUT_FILE)?;
    for line in BufReader::new(file).lines() {
        let line = line.unwrap();
        for cap in regex.captures_iter(&line) {
            persons.insert(cap[1].to_string());
            persons.insert(cap[4].to_string());
            let mut h: isize = cap[3].parse()?;
            match &cap[2] {
                "gain" => (),
                "lose" => h *= -1,
                _ => panic!("Invalid option")
            }
            happiness.insert((cap[1].to_string(), cap[4].to_string()), h);
        }
    }
    let mut persons: Vec<String> = persons.into_iter().collect();

    // --- Puzzle 1 ---
    let max = calculate_happiness(&persons, &happiness);
    println!("1. Total change in happiness: {}", max);

    // --- Puzzle 2 ---
    persons.push("me".to_string());
    for p in &persons {
        happiness.insert((p.to_string(), "me".to_string()), 0);
        happiness.insert(("me".to_string(), p.to_string()), 0);
    }
    let max = calculate_happiness(&persons, &happiness);
    println!("2. Total change in happiness: {}", max);

    Ok(())
}

File: 2015/13/go/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"math"
	"os"
	"regexp"
	"slices"
	"strconv"
	"strings"
)

// --- Constants ---
const FILENAME = "../input/input.txt"
const REGEX = `(.+)? would (gain|lose) (\d+) happiness units by sitting next to (.+)\.`

// Calculates the permutations
func permutations(a []string) [][]string {
	var r [][]string

	var generate func(int)
	generate = func(i int) {
		if i == len(a) {
			tmp := make([]string, len(a))
			copy(tmp, a)
			r = append(r, tmp)
			return
		}
		for j := i; j < len(a); j++ {
			a[i], a[j] = a[j], a[i]
			generate(i + 1)
			a[i], a[j] = a[j], a[i]
		}
	}
	generate(0)
	return r
}

// Calculates the happiness
func calculateHappiness(persons []string, happines map[string]int) int {
	mx := math.MinInt
	for _, p := range permutations(persons) {
		h := 0
		for i := 0; i < len(p); i++ {
			n1 := p[i]
			n2 := p[(i+1)%len(p)]
			h += happines[n1+"-"+n2]
			h += happines[n2+"-"+n1]
		}
		mx = max(mx, h)
	}
	return mx
}

func main() {
	// --- Variables ---
	re := regexp.MustCompile(REGEX)
	persons := []string{}
	happiness := make(map[string]int)

	// --- Read input file ---
	file, err := os.Open(FILENAME)
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()

	// --- Parse each line ---
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		m := re.FindStringSubmatch(line)
		if !slices.Contains(persons, m[1]) {
			persons = append(persons, m[1])
		}
		if !slices.Contains(persons, m[4]) {
			persons = append(persons, m[4])
		}
		h, err := strconv.Atoi(m[3])
		if err != nil {
			log.Panic(err)
		}
		if m[2] == "lose" {
			h *= -1
		}
		happiness[m[1]+"-"+m[4]] = h
	}

	// --- Puzzle 1 ---
	mx := calculateHappiness(persons, happiness)
	fmt.Println("1. Max change in happiness:", mx)

	// --- Puzzle 2 ---
	persons = append(persons, "me")
	for _, p := range persons {
		happiness[p+"-me"] = 0
		happiness["me-"+p] = 0
	}
	mx = calculateHappiness(persons, happiness)
	fmt.Println("2. Max change in happiness:", mx)
}
